Defer instrumentation of Redis connection

DamianEdwards opened the pull request:
Fixes #2907
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2982)

DamianEdwards committed:
Defer adding Redis connection to instrumentation until it's requested

DamianEdwards committed:
Defer instrumentation of Redis connection until created

Fixes #2907

davidfowl commented on src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:
```diff
@@ -8,6 +8,9 @@
 using Microsoft.Extensions.Options;
+using OpenTelemetry.Instrumentation.StackExchangeRedis;
+
+//using OpenTelemetry.Instrumentation.StackExchangeRedis;
```
Dead code

davidfowl commented on src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:
```diff
@@ -133,33 +136,37 @@ public static class AspireRedisExtensions
-                });
+                    .WithTracing(t =>
+                    {
+                        t.AddSource(typeof(StackExchangeRedisInstrumentation).Assembly.GetName().Name!);
```
Just hardcode the source name. Store it in a constant.

DamianEdwards committed:
PR feedback

DamianEdwards commented:
The test failure is odd. I looked through it with @davidfowl and it appeared more like a race condition somewhere in our test code or OpenTelemetry itself and I haven't been able to repro it locally (logged #2987). Open to other ideas.

DamianEdwards committed:
Test fix

eerhardt commented on tests/Aspire.Components.Common.Tests/ActivityNotifier.cs:
```diff
@@ -20,6 +20,6 @@ public sealed class ActivityNotifier : BaseProcessor<Activity>
     {
         ExportedActivities.Add(data);
-        _taskSource.SetResult();
+        _taskSource.TrySetResult();
```
This seems to indicate that OnEnd is getting called multiple times now. (But it wasn't before?)

This may be an issue.

eerhardt replied:
These tests should be able to be executed and debugged on your local machine. You just need Docker running so the TestContainers works.

DamianEdwards replied:
I haven't been able to reproduce the failure we're seeing in Helix locally. I debugged through this with @davidfowl and the `ActivityNotifier` isn't even explicitly added by the tests that are failing but this was something we saw that could conceivably contribute if tests are running in parallel.

DamianEdwards replied:
Looks like Linux passed this time.

DamianEdwards replied:
Sooo, looks like this change fixed it.

DamianEdwards replied:
OK we figured this out, it's an actual issue with the code change to do with a workaround to support keyed services in the Redis OTel client. Fixing.

eerhardt replied:
Doing more investigation, this was caused by a test that registered an ActivityListener in the same process as the rest of the tests. ActivityListeners are global, so they get all Activities. The Activities were never created before because the Host was never Started. With the latest changes the Activities always get created even if the Host isn't Started.

The fix is to move the ActivityListener  tests to a separate process.

eerhardt committed:
Fix tests to work correctly with the new tracing implementation.

- Tests that register an ActivityListener need to run in a separate process so they don't get Activities from other tests.
- ActivityNotifier needs to run its Task continuations asynchronously, or tests can deadlock when continuing on the "drain" background thread.

Also remove dead code.

eerhardt commented on src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:
```diff
@@ -133,33 +136,36 @@ public static class AspireRedisExtensions
         {
             // Supports distributed tracing
+            // We don't call AddRedisInstrumentation() here as it results in the TelemetryHostedService trying to resolve & connect to IConnectionMultiplexer
+            // via DI on startup which, if Redis is unavailable, can result in an app crash. Instead we add the ActivitySource manually and call
+            // ConfigureRedisInstrumentation() to ensure .
```
@DamianEdwards - this comment trails off.

eerhardt committed:
PR feedback

DamianEdwards committed:
Ensure AddInstrumentation is called for Redis

DamianEdwards commented on src/Components/Aspire.StackExchange.Redis/AspireRedisExtensions.cs:
```diff
@@ -133,34 +136,29 @@ public static class AspireRedisExtensions
+                    // This ensures the core Redis instrumentation services from OpenTelemetry.Instrumentation.StackExchangeRedis are added
+                    t.ConfigureRedisInstrumentation(_ => { });
+                    // This ensures that any logic performed by the AddInstrumentation method is executed (this is usually called by AddRedisInstrumentation())
+                    t.AddInstrumentation(sp => sp.GetRequiredService<StackExchangeRedisInstrumentation>());
```
It felt "safer" to me to ensure we're calling all the methods that `AddRedisInstrumentation()` calls under the hoods now that we're not calling it.

eerhardt replied:
OK, that seems reasonable.

eerhardt commented:
/azp run
